### DISCUSSION TITLE
Assembling matrix for coupled thermoelastic problem

### DISCUSSION THREAD

**Discussion by malana002** [kinnala/scikit-fem#1028] [Q&A]
Hello,

I am trying to obtain the matrices to solve a coupled thermoelastic problem, and for that I need to assemble the $K^{uT}$ matrix from the following expression:
![image](https://user-images.githubusercontent.com/132545956/236132029-eaeb9323-8323-483b-836f-f825240d0d6a.png)

$K^{uT}$ is the matrix that couples the thermal and elastic problem, and is defined as:
![image](https://user-images.githubusercontent.com/132545956/236138772-4aa1e1d9-5965-47dd-ba4c-7f90cdd13015.png)

$\beta$ is defined as $\beta = [D]\left(\alpha\right)$, where $[D]$ corresponds to the elastic stiffness matrix and $\alpha = [\alpha_x, \alpha_y, \alpha_z, 0, 0, 0]^T$ represents the heat expansion coefficient vector. 

How could I obtain the $K^{uT}$ matrix for ElementTetP1 and ElementTetP2 element basis?

**Comment by kinnala** [comment]
Unfortunately it's not clear to me what are $B$ and $N$ in this case. It might help to write down the variational formulation of the coupled problem.

The text you quote seems to be written with a lower level of abstraction in mind since it already talks about the stiffness matrices derived from the variational formulation.

**Comment by kinnala** [comment]
It might help to think about a simpler 1D example first. I tried reading some references on 1D thermoelasticity.

The governing equation for a static bar could be
$$-(EA(u' - \alpha T))' = f.$$
The temperature $T$ could be governed by a static heat equation:
$$-(kT')' = q.$$

The weak formulation could be
$$(EAu', v') + (\alpha T', v) = (f,v)$$
and
$$(kT', V') = (q, V).$$

This problem can be solved like this:
```python
from skfem import *
from skfem.helpers import *

m = MeshLine().refined(5)
basis = Basis(m, ElementLineP1() * ElementLineP1())

@BilinearForm
def bilinf(u, T, v, V, w):
    return dot(grad(u), grad(v)) + 0.1 * grad(T)[0] * v + dot(grad(T), grad(V))

@LinearForm
def linf(v, V, w):
    # heating between (0.3, 0.7)
    return 0. * v + (w.x[0] > 0.3) * (w.x[0] < 0.7) * V

A = bilinf.assemble(basis)
b = linf.assemble(basis)

y = solve(*condense(A, b, D=basis.get_dofs().all()))

(u, ubasis), (T, Tbasis) = basis.split(y)

ubasis.plot(u).show()
```

If you just need the coupling matrix, you can do something like:
```python
ubasis = Basis(m, ElementLineP1())
Tbasis = Basis(m, ElementLineP1())

@BilinearForm
def couplform(T, v, w):
    return 0.1 * grad(T)[0] * v

C = couplform.assemble(Tbasis, ubasis)
```
![displacement](https://user-images.githubusercontent.com/973268/236678152-5189dfd0-f078-42a0-98ac-33f6daa75e11.png)

**Comment by malana002** [comment]
Thank you so much for the help, I really appreciate it. In my case, I would like to solve it for 3D space using ElementTetP1 and ElementTetP2 elements. Thus, following your example just for the coupled matrix, how would it work for these bases?
```python
ubasis = Basis(m, ElementVector(ElementLineP1()))
Tbasis = Basis(m, ElementLineP1())
```
